File: 0279-perfect-squares/0279-perfect-squares.py
class Solution:
    def numSquares(self, n: int) -> int:
        dp=[0 for i in range(n+1)]
        dp[1]=1
        sq=[]
        i=1
        while(i*i<=n):
            sq.append(i*i)
            i+=1
        for i in range(2,n+1):
            temp=math.inf
            for x in sq:
                if(x<=i):
                    temp=min(temp,dp[i-x])
                else:
                    break
            dp[i]=temp+1
        # print(dp)
        return dp[n]
            
                
                        
                        
        
            
        

File: 0279-perfect-squares/0279-perfect-squares.cpp
class Solution {
private:
    int solve(int n,vector<int>&Sq,vector<int>&DP)
    {
        if(n==0)return 0;
        else
        {
            if(DP[n]!=-1)return DP[n];
            int ans=INT_MAX;
            for(auto x:Sq)
            {
                if(x<=n)
                    ans=min(ans,solve(n-x,Sq,DP));
            }
            return DP[n]=ans+1;
        }
    }
public:
    int numSquares(int n) {
        vector<int>Sq;
        vector<int>DP(n+1,-1);
        int i=1;
        while(i*i<=n)
        {
            Sq.push_back(i*i);
            i++;
        }
        return solve(n,Sq,DP);
    }
};